### ISSUE TITLE
--link should be disallowed with --net=container

### ISSUE THREAD

**Issue by gdm85** [docker/docker#9340]
I am trying to understand and use the ``--net=container`` feature. When I use a command like this:
```
docker run -d --restart=no --cpu-shares=512 --link db:db -p 443 --net=container:other-web-app -v /etc/someVolume  --name=test-net-container base-image
```
I get this error:
```
2014/11/25 21:51:43 Error response from daemon: Cannot start container 1e02f12a2fdceb2f057f0cd58cf29ffe45790dc8ecae7be88456f79dccc58c7e: parent IP '' is invalid
```

I have not been able to find where this error is coming from (doesn't seem docker/libcontainer code), however it is my guess that you cannot link to a container when using ``--net=container`` mode. I got there by successfully running the same command without the ``--link`` option :
```
docker run -d --restart=no --cpu-shares=512 -p 443 --net=container:other-web-app -v /etc/someVolume  --name=test-net-container base-image
```

If I am not mistaken, shouldn't the ``--link`` option be disallowed (or at least a more explanatory error message given)?

**Comment by AlexAndrascu** [comment]
Confirmed!
Also getting this in 1.4.1:

```
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 5bc2ff8
```

**Comment by gdm85** [comment]
@LK4D4 can we make a test for this issue?

**Comment by sirwolfgang** [comment]
I am running into this with 1.5.0

**Comment by willhf** [comment]
#dibs

**Comment by tiborvass** [comment]
Ping @cpuguy83 isn't this fixed by #11369 ?

**Comment by coolljt0725** [comment]
@tiborvass  #11369   allow `--link` work with `--net=container`. But may be disallow `--link` work with
`--net=container` is a desired behavior, see https://github.com/docker/docker/blob/master/runconfig/parse.go#L20
But this line never work due to the wrong if logic in https://github.com/docker/docker/blob/master/runconfig/parse.go#L128
(and I don't realize this when I work on #11369 and thought allow `--link` work with `--net=container` is a desired behavior at that time, sorry for this)
#11516 are related to this

**Comment by cpuguy83** [comment]
@tiborvass It is not, I actually was working with @willhf on this issue and tested this, got a pretty bad error for it, even on master.

**Comment by estesp** [comment]
Just a comment for posterity, given it may be confusing from prior comments across the various issues and PRs:

What @coolljt0725 fixed in #11369 allows the use of `--link` where the link's target container actually was started with `--net=container:<some_running_container>`, and it even supports that recursively (if that target points its net to another running container which has it's net pointing to another container).  So, that fixed a valid problem, but what appears to still an issue (fixed at least by correcting the flag conflict check in #11686) is trying to start a container whose net is a pointer to another container, **and** link to yet another container at the same time.  This seems reasonable to block until we determine if that can be fixed (which would require adding the link target's IP into the parent net's container `/etc/hosts`, meaning we have linked the `--net=container:<other_container>` to the target as well, which may or may not be appropriate and needs some thought.
